### ISSUE TITLE
Infinite loop of specific file

### ISSUE THREAD

**Issue by mihai-stancu** [composer-unused/composer-unused#659]
## Infinite loop on specific file
On project which included Zend 1.x components `composer-unused` was hanging forever.

I tried to debug and noticed that it entered an infinite loop on a specific file.

## Additional information
~~I tried to run `composer-unused` on a small subset of files (just the directory with the Zend libraries) but that didn't trigger the infinite loop.~~ (I was wrong, my setup on this was not correct, I can reproduce it on a smaller subset)

Before entering the loop a Notice was issued:
```
PHP Notice:  file_get_contents(): Read of 12288 bytes failed with errno=21 Is a directory in /opt/composer-unused/vendor/composer-unused/symbol-parser/src/File/FileContentProvider.php on line 21
```

To identify the loop I edited the file `FileContentProvider` and printed the filename -- which showed all the files until the infinite loop kicked in and that specific file was printed forever. Removing that file resolves the issue.

I am trying to isolate the problem to document a clear reproducible usecase.

Note: `composer-unused` is installed as a separate directory in `/opt/composer-unused` via composer install.

**Comment by mihai-stancu** [comment]
I can confirm that one single file is the cause and if that file is present in an otherwise empty project it still triggers the infinite loop.

### Setup:
```bash
mkdir test;
cd test;
cp offensive_file.php ./
cp composer.json ./
composer install;
composer-unused;
```

```json
// composer.json
{
    "name": "test/test",
    "autoload": {"classmap": ["."]}
}
```

The file is something from (some old version of) Zend 1.x. I'll try to isolate the exact lines that cause the issue otherwise I'll post the entire file.

**Comment by mihai-stancu** [comment]
Here's the infinit loop generator (minimal file):
```php
// Test.php -- name of file must match the require_once path
<?php

class Test
{
    public static function test()
    {
        require_once ('Test.php'); // original file had require_once ($someConfigPath . 'Test.php');
    }
}
```

**Comment by mihai-stancu** [comment]
In that Zend component the offending file was an addapter file:

```
// path is Zend/Some/Namespace/Adapter/Thing.php
class Zend_Some_Namespace_Adapter_Thing {
    public static function test()
    {
      if (!class_exists('Thing')) { // this is a different class in  a different directory, happens to have same file name
          require_once ($configPath . 'Thing.php');
      }
    }
}
```

**Comment by mihai-stancu** [comment]
I'll attempt to identify where the loop comes from in the `composer-unused` package.

**Comment by mihai-stancu** [comment]
@icanhazstring I wasn't able to grok the dataflow thru the symbol parser / iterators etc. so I can't pinpoint the best point in the code to stop the loop.

However I saw the other 2 issues posted by others earlier this year so I guess it's not such a weird corner-case I stumbled into.

I'm sure we can fix this with a CachedFileContentProvider that decorates the FileContentProvider. I'll write a PR for that.

**Comment by icanhazstring** [comment]
Yes probably. There is a path in the code, which follows `include`, `include_once`, `require` and `require_once`.

So it would be best to write a test where two files, include each other. Meaning it should not break but complete the test.

So caching included files would be probably the way to go.

**Comment by mihai-stancu** [comment]
Correction not a CachedFileContentProvider but a UniqueFileContentProvider. Since we're not returning the cached content, we're returning an empty string if the file was already parsed.

Do you have a preference on how I implement this? 2 separate files (decortator + original) vs. just modifying the FileContentProvider (in a way that keeps backwards compat).

**Comment by icanhazstring** [comment]
> Do you have a preference on how I implement this? 2 separate files (decortator + original) vs. just modifying the FileContentProvider (in a way that keeps backwards compat).

As this is clearly a bug. I think changing the behavior should still keep backward compat.
So maybe keep the change in the FileContentProvider.

**Comment by mihai-stancu** [comment]
I found a better approach: `AbstractCollector::followIncludes` makes more sense to stop the production of the recursion than to circumvent it.

This was the call site I was looking for previously but only found it after writing a Unit test.

**Comment by icanhazstring** [comment]
Should be solved with latest release.
Closing it. If the error comes up again, please open a new issue.
